scylla/src/transport/speculative_execution.rs

```diff
@@ -0,0 +1,57 @@
+use futures::future::FutureExt;
+use futures::stream::{FuturesUnordered, StreamExt};
+use std::future::Future;
+use std::time::Duration;
+
+use super::errors::QueryError;
+
+#[derive(Debug, Clone)]
+pub struct SpeculativeExecutionPolicy {
+    pub attempts_count: usize,
+    pub delay: Duration,
+}
+
+impl SpeculativeExecutionPolicy {
+    pub async fn execute<QueryFut, ResT>(
+        &self,
+        query_runner_generator: impl Fn() -> QueryFut,
+    ) -> Result<ResT, QueryError>
+    where
+        QueryFut: Future<Output = Option<Result<ResT, QueryError>>>,
+    {
+        let mut available_attempts = self.attempts_count;
+
+        let mut async_tasks = FuturesUnordered::new();
+        async_tasks.push(query_runner_generator());
+
+        let sleep = tokio::time::sleep(self.delay).fuse();
+        tokio::pin!(sleep);
+
+        loop {
+            futures::select! {
+                _ = &mut sleep => {
+                    if available_attempts > 0 {
+                        async_tasks.push(query_runner_generator());
+                        available_attempts -= 1;
+
+                        // reset the timeout
+                        sleep.set(tokio::time::sleep(self.delay).fuse());
+                    }
+                }
+                res = async_tasks.select_next_some() => {
+                    match res {
+                        Some(r) => return r,
```

Comment:
I think we need to ignore the error here (or at least some kinds of errors) if there are some in-flight executions.

Consider a scenario with the following events, in this order:

- Request is sent to node 1
- We speculatively send request to node 2
- Node 1 times out and returns an error - this will cause us to return an error here
- Node 2 responds with a success - but we ignore it here

Reply:
Done, i chose to ignore IoError and TimeoutError.